linkerd/cache/src/track.rs

```diff
@@ -1,33 +1,63 @@
 use crate::Handle;
 use linkerd2_stack::{layer, NewService};
-use std::task::{Context, Poll};
+use std::{
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::{sync::Notify, time};
+use tracing::{debug, debug_span, trace};
+use tracing_futures::Instrument;
 
 #[derive(Clone)]
 pub struct NewTrack<N> {
     inner: N,
+    idle: time::Duration,
 }
 
 #[derive(Clone)]
 pub struct Track<T> {
     inner: T,
-    _handle: Handle,
+    notify: Arc<Notify>,
 }
 
-/// === impl NewTrack ===
-
 impl<N> NewTrack<N> {
-    pub fn layer() -> impl layer::Layer<N, Service = NewTrack<N>> + Copy + Clone {
-        layer::mk(|inner| NewTrack { inner })
+    pub fn layer(
+        idle: time::Duration,
+    ) -> impl layer::Layer<N, Service = NewTrack<N>> + Copy + Clone {
+        layer::mk(move |inner| Self { inner, idle })
     }
 }
 
 impl<T, N: NewService<T>> NewService<(Handle, T)> for NewTrack<N> {
     type Service = Track<N::Service>;
 
-    #[inline]
-    fn new_service(&mut self, (_handle, target): (Handle, T)) -> Self::Service {
+    fn new_service(&mut self, (handle, target): (Handle, T)) -> Self::Service {
+        let idle = self.idle;
+        let notify = Arc::new(Notify::new());
+
+        // Create a background task that drops the cache entry's handle when the
+        // service hasn't been notified for an `idle` interval.
+        tokio::spawn({
+            let notify = notify.clone();
+            async move {
+                loop {
+                    tokio::select! {
+                        _ = notify.notified() => {
+                            trace!(idle = idle.as_secs(), "Resetting");
+                        }
+                        _ = time::sleep(idle) => {
+                            debug!(idle = idle.as_secs(), "Idled out");
+                            drop(handle);
+                            return;
+                        }
+                    }
+                }
+            }
+            .instrument(debug_span!("tracker"))
```

Comment:
nit/tioli: _could_ just stick the idle secs in the span:
```suggestion
                            trace!("Resetting");
                        }
                        _ = time::sleep(idle) => {
                            debug!("Idled out");
                            drop(handle);
                            return;
                        }
                    }
                }
            }
            .instrument(debug_span!("tracker", idle = idle.as_secs()))
```

Reply:
nit: incidentally, also, right now this will look like this, supposing the duration is ten seconds:
```
idle=10
```
this doesn't indicate that it's in seconds.

I would either rename the field to `idle_secs` or something, or just use `Duration`'s `fmt::Debug` impl, [which now pretty-prints the duration with a unit](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=22b1d21167f7acd102e7006f942731f1).